### ISSUE TITLE
Integration of LibreFOMO (Faster Objects, More Objects)

### ISSUE THREAD

**Issue by bencejdanko** [Libre-YOLO/libreyolo#300]
FOMO is optimized for constrained edge, NPU, and microcontroller environments. FOMO treats object detection as a dense spatial classification task over a low-resolution grid. Currently, I see only that [Edge Impulse](https://docs.edgeimpulse.com/studio/projects/learning-blocks/blocks/object-detection/fomo) supports an officially supported training scheme for FOMO models on their platform.

* **Model Family Name**: `LibreFOMO`
* **Internal Identifier**: `FAMILY = "librefomo"`
* **Filename Prefix**: `FILENAME_PREFIX = "LibreFOMO"`

### 1. Backbone and Dynamic Resolutions

The backbone network slices a pretrained MobileNetV2. Capacity and input size vary across implementations:

| Variant | Input Resolution | Width Multiplier ($\alpha$) | Target Feature Map / Grid Size |
| --- | --- | --- | --- |
| **Small (`s`)** | $96 \times 96$ | 0.35 | $12 \times 12$ |
| **Medium (`m`)** | $192 \times 192$ | 0.50 | $24 \times 24$ |
| **Large (`l`)** | $256 \times 256$ | 1.00 | $32 \times 32$ |

> [!IMPORTANT]
> By default, `torchvision.models.mobilenet_v2` does not contain weights for MobileNetV2 with alpha < 1.0 (channel width multiplier, which scales model size and inference time significantly). We need to point to distinct checkpoint files and match expected dimensions in the abstraction.

### 2. Dense Grid Classification Head

A $1 \times 1$ 2D Convolutional layer maps the feature channels directly to the number of target classes plus an explicit background channel:

```python
import torch.nn as nn

# Head implementation mapping features to classes
self.head = nn.Conv2d(feat_channels, num_classes + 1, kernel_size=1)

```

Channel `0` is designated as the explicit background class. A spatial softmax is applied over the channel dimension per grid cell. Each cell acts as an independent, mutually exclusive classifier for the background class and target class.

### 3. Target Construction & Weighted Cross-Entropy Loss

During training, target bounding box centroids are mapped to discrete grid cell coordinates $(i, j)$. Cell $(i, j)$ is assigned the target class index of the overlapping object. Unassigned cells default to `0` (background).  combat heavy background imbalance, we apply a high foreground weight multiplier ($W_{\text{obj}} = 100.0$) relative to background ($W_{\text{bg}} = 1.0$) inside a standard Cross-Entropy loss.

### 4. Post-processing & Inference Pipeline

1. **Thresholding**: Cells with a foreground class probability above a configurable confidence threshold are selected.
2. **Local Maximum Suppression (LMS)**: A local max-pooling filter over the spatial probabilities suppresses adjacent activations firing on the same object prior to extraction.
3. **Grid-to-Centroid Reconstruction**: Active cells are converted back to normalized bounding boxes centered at the cell's center with a predefined size matching the cell's receptive field.

The model returns the standard LibreYOLO dictionary format: `{"boxes", "scores", "classes", "num_detections"}`.

## Challenges

The current metrics for object detection uses Intersection-over-Union (IoU). FOMO predicts fixed-size centroids and box sizes (relative to the input image), which will cause IoU metrics to be artificially low. We have to implement custom metrics calculation for FOMO, which would be done by measuring precision and recall on the grid classification level.

## Expected Codebase Changes

```
libreyolo/
├── models/
│   └── librefomo/                 [NEW]
│       ├── __init__.py            [NEW] Exports LibreFOMO
│       ├── model.py               [NEW] BaseModel subclass implementing lifecycle
│       ├── nn.py                  [NEW] MobileNetV2 backbone + 1x1 head configuration
│       ├── loss.py                [NEW] Grid-builder target matching & weighted CE loss
│       └── utils.py               [NEW] Centroid postprocessing & evaluation helpers
├── models/__init__.py             [MODIFY] Register LibreFOMO import
├── __init__.py                    [MODIFY] Add LibreFOMO to top-level exports
├── training/config.py             [MODIFY] Add FOMOConfig validation block
├── validation/
│   └── preprocessors.py           [MODIFY] Add FOMOValPreprocessor
tests/
└── unit/
    └── test_librefomo.py          [NEW] Comprehensive unit test suite

```

---

### Execution Commands

```bash
# Run all unit tests for LibreFOMO
pytest tests/unit/test_librefomo.py -v

# Run only build and forward shape tests
pytest tests/unit/test_librefomo.py -k "test_build_and_forward" -v

# Run loss target and backward differentiation tests
pytest tests/unit/test_librefomo.py -k "test_loss_forward" -v

# Verify model registry loading behaviors
pytest tests/unit/test_librefomo.py -k "can_load" -v

# Verify factory registration doesn't break existing models
pytest tests/unit/test_factory.py tests/unit/test_model_factory.py -v

```

## Implementation Roadmap

* [ ] Implement backbone slicing and width scaling configurations in `models/librefomo/nn.py`.
* [ ] Create the target tensor generator and weighted cross-entropy loss in `loss.py`.
* [ ] Set up post-processing with Local Maximum Suppression (LMS) and raw export options.
* [ ] Integrate registry functions (`can_load`, state-dict introspection) in `model.py`.
* [ ] Implement standalone centroid localization evaluation metrics.

**Comment by EHxuban11** [comment]
Hello @bencejdanko , thank you for opening the issue, adding FOMO is a very promising idea.

Regarding the backbone I've searched the code and weights just so that we have them here in the issue:

_Code (apache2) :_
- Code: https://github.com/keras-team/keras/blob/master/keras/src/applications/mobilenet_v2.py
- License: https://github.com/keras-team/keras/blob/master/LICENSE

_Here are the backbones:_
- Small:  https://storage.googleapis.com/tensorflow/keras-applications/mobilenet_v2/mobilenet_v2_weights_tf_dim_ordering_tf_kernels_0.35_96_no_top.h5
-  Medium: https://storage.googleapis.com/tensorflow/keras-applications/mobilenet_v2/mobilenet_v2_weights_tf_dim_ordering_tf_kernels_0.5_192_no_top.h5
- Large: https://storage.googleapis.com/tensorflow/keras-applications/mobilenet_v2/mobilenet_v2_weights_tf_dim_ordering_tf_kernels_1.0_224_no_top.h5

In LibreYOLO and most similar libraries every model is pretrained with a series of classes, usually COCO. In the case of LibreFOMO it could be just the class person, for example, but I'm open to ideas, what are you imagining?

For adding a new model to libreyolo, there is this skill that could help if you use coding agents: https://github.com/LibreYOLO/libreyolo/blob/dev/skills/libreyolo-port-model/SKILL.md

For validation FOMO probably needs it's own validator FOMOValidator. Publicly it can still be a detect model, but internally the validator should report centroid/grid precision and recall instead of the default IoU-based COCO metrics. BaseModel should not be overriden, it should be subclassed normally as you say here in the issue.

**Comment by bencejdanko** [comment]
@EHxuban11 

Thank you very much for sourcing the exact the weights. I will convert these to a PyTorch implementation for LibreFOMO small, med, large versions.

I believe a 1 person class focus would be correct for the time being. I find FOMO works quite well on this particular overhead dataset: https://loafisheye.github.io/. 

I'll train it from this dataset source:
https://huggingface.co/datasets/bdanko/loaf_resolution_512. 

We will predict just 1 class, person.

I will find the best way to approach this. I'll avoid overriding the BaseModel, and add a new "Point" task (alongside Detection, Segmentation, and Pose). By Point we imply single-coordinate object locations. I suggest we use PointValidator, Predictions will look like `[(x_pred, y_pred, class, confidence), ...]`. Metrics would be precision, recall, F1, and mean distance error.

**Comment by bencejdanko** [comment]
An update, I have converted the keras backbones to pytorch tensors and architecture. Temporarily hosted at https://huggingface.co/bdanko/LibreFOMO (conversion script at https://github.com/bencejdanko/convert_keras_mobilenetv2_to_pytorch)

I will be continuing to validate that these can be properly trained and exported for people detection, and continue design for a "point" task and PointValidator class.

**Comment by EHxuban11** [comment]
Hello @bencejdanko,  thank you for the work.

I've been a bit busy with the release, the LibreYOLO hackathon and other things, sorry for taking so much time to answer.

I went through the issue and PR and it's clearly solid. You closed the PR because you felt like you overextended by adding a bunch of code related to a new point task and so on.  What I've done is to add  the task point myself to dev exactly like the other tasks and documented it so that you don't have to worry about that. Now this task model can suport LibreFOMO and maybe in the future CenterNet, CSRNet, etc.

The important documentation for the point task are in:
docs/adr/0003-point-task-contract.md, docs/nomenclature.md, docs/checkpoint_schema.md, docs/dataset_schema.md and libreyolo/tasks.py.

If you're still up for it, please open a new PR against the up to date dev which has the point task merged.

Once that is done we can proceed with training models if necessary, uploading them to the libreYOLO HF so that people can get them, evaluating and publishing the evaluations in visionanalysis.org.

Thank you again for your work

**Comment by bencejdanko** [comment]
Got it, I'll try and break this down into better steps so it will integrate better. Next PR will be PointValidator with these metrics:

To evaluate how accurately (x,y) coords are predicted:

- Precision, recall, f1-score at different distance thresholds
- Average precision
- Mean localization error
- mAP@d (mean average precision at distance d)
- mAP@d [0.01:0.10] (mean average precision over range)

Then for density and counting accuracies:

- Mean absolute error
- Root mean squared error

I'll create tests for these in `tests/unit/test_point_validator.py`.

**Comment by EHxuban11** [comment]
Hello @bencejdanko , you can do one single PR if you want.

**Comment by testdummyvt** [comment]
@bencejdanko This is a pretty interesting model and like the idea of just pointing. I am trying to find more resources on this but I only found EdgeImpulse writeup: https://docs.edgeimpulse.com/tutorials/end-to-end/object-detection-centroids

Are there any other publications on this method? articles, papers, blogs you would suggest.

**Comment by bencejdanko** [comment]
> [@bencejdanko](https://github.com/bencejdanko) This is a pretty interesting model and like the idea of just pointing. I am trying to find more resources on this but I only found EdgeImpulse writeup: https://docs.edgeimpulse.com/tutorials/end-to-end/object-detection-centroids
> 
> Are there any other publications on this method? articles, papers, blogs you would suggest.

I actually only found the FOMO model as well through the edge impulse documentation as well, which I thought is why it might be a good area of research and create an openly available solution.

If you would like to try it out, I built a library around it with an example at https://github.com/fomo-edge-ai/fomo/tree/master/examples, where I use it to train on a small collected dataset. After optimizing exports we can get models in the tens of kbs, to deploy on microcontrollers, which is why I was so interested, but I still need to improve the technique.

For readup, I would suggest MobileNetV2 / depthwise convolutions, since that is our feature extractor and limitation. 

There are also more computationally expensive methods like CenterNet which have complicated point extractions, e.g they extract points, but also predict the space around the point, which is something to think about for the future. NVIDIA’s new LocateAnything model I believe has a pointer head, though I’m not sure the details. 

On FOMO specifically, I have not found anything else, though it may be such models go by different names in some papers since it is conceptually simple.

**Comment by bencejdanko** [comment]
An update on what I am planning:

Currently, I don't believe FOMO would be worth much continue implementing for LibreYOLO. I believe it's out of of the scope for what LibreYOLO should maintain:

- Since FOMO is an incredibly small model, it does not have the capacity to generalize and perform well out of the box
- It is a more practical model to attune to very specific environments, but not a great candidate for hosting generalized weights

Instead, I am thinking of training a regular LibreYOLO9 backbone, in which we simply swap out the head to predict our [x, y, class, confidence] target directly. It may be something like "LibreYOLO9t-point".

I am planning to pre-train on [Open Images v7's point-based annotations](https://wandb.ai/telidavies/ml-news/reports/Open-Images-V7-Released-Adds-New-Point-Based-Labeling-Visualizers--VmlldzoyODU4NjU2) (Creative Commons license). 

One nuance of this dataset is that each object is covered by many point annotations for the same class (see below an Open Image point annotation sample). So it's not great for say, exact and precise counting tasks.

For very precise counting one would need to continue training on datasets where each object is only labeled once. 

However, for LibreYOLO9-Point, I think we could simply offer weights that are pre-trained off of Open Images v7 point-level annotations, and leave the fine-tuning for counts up to end users. Having such pre-trained weights would be a great generalized starting point for end users 

<img width="1529" height="715" alt="Image" src="https://github.com/user-attachments/assets/c5dc4391-cac2-4177-a285-c9a1b5d9736c" />

**Comment by EHxuban11** [comment]
Hello @bencejdanko , 

I hope you are well!

I think that the original FOMO does not generalize and perform well out of the box, but it does work fine for some specific use cases. As long as LibreYOLO's FOMO implementation is as good os better as the original one it whould be fine. It makes me think about pretrained weights, maybe we can just think about distributing untouched mobilenet backbone with random initialized weights for the rest of the network for the folks who want to fine tune. We can also train some models as a showcase, like people detection and some industrial use-case such as counting bolts or whatever.

Regarding the implementation of a new point model based on yolo9, I'm not against it if it works XD
What worries me a bit is that mobilenet is very edge-firendly with the operations it uses while the yolo9 backbone is a questionmark. 

Thanks for the links, very interesting.

I would like to work on LibreFOMO to experiment with it. Maybe I arrive to the same conclusions as you. You had a repo, right? 

Thanks for your work as always

**Comment by bencejdanko** [comment]
@EHxuban11 

This sounds solid, in that case, I will continue trying to improve it. 

For one case, I picked out a conveyer belt + garlic dataset from Roboflow. My current LibreFOMO attempt only reaches ~0.83 F1:

https://github.com/fomo-edge-ai/fomo/blob/master/examples/train_fomo_garlic_conveyer%3D%3D0.0.12.ipynb

<img width="1790" height="539" alt="Image" src="https://github.com/user-attachments/assets/536fb392-e784-4aed-b41a-1d84781cb969" />

To compare performance, we can just use the Edge Impulse Studio:

https://studio.edgeimpulse.com/

Within the studio, I get:

0.99 non-background precision
0.87 non-background recall
0.92 non-background f1
95% accuracy (defined as percentage test samples with precision score above 80%)

I will start a PR to get what I currently have integrated, then we can continue to try and find ways to improve training. It may be that Edge Impulse has better feature extractor, or better post processing calculation to calculate centroids, I will explore. We can do these applications:

- People detection
- Car counting in parking lots
- conveyer belt application
